src/controller/java/AndroidCallbacks.cpp

```diff
@@ -165,8 +166,15 @@ ReportCallback::~ReportCallback()
         env->DeleteGlobalRef(mSubscriptionEstablishedCallbackRef);
     }
     env->DeleteGlobalRef(mReportCallbackRef);
-    if (mReadClient != nullptr)
+    if (app::InteractionModelEngine::GetInstance()->InActiveReadClientList(mReadClient))
     {
+        ChipLogError(Controller, "shutdown sub");
+        if (mReadClient->IsSubscriptionType() && mReadClient->GetSubscriptionId().HasValue())
+        {
+            app::InteractionModelEngine::GetInstance()->ShutdownSubscription(
```

Comment:
yes, i think we don't need it. But currently destructor for ReadClient does not automatically shutdown subscription.
1. user needs explicitly call ShutdownSubscription, which further call ReadClient::Close, then destroy client in onDone
2. when fabric is removed, it trigger ReadClient::Close, then destroy client in onDone.
3. when Response timeout or liveness timeout happens,  it trigger ReadClient::Close, then destroy client in onDone.
I think the expose for ShutdownSubscription API to Java controller should be enough in this PR.

Reply:
>  But currently destructor for ReadClient does not automatically shutdown subscription.

"Shut down subscription" just calls `Close(CHIP_NO_ERROR)` on the `ReadClient`.  

`Close` does the following things:

1. `ClearActiveSubscriptionState` (which just sets some state)
2. `StopResubscription` (cancels timers)
3. `OnDone`

OnDone then typically deletes the ReadClient.

The ReadClient destructor, if it's a subscription client, calls `StopResubscription` and calls `RemoveReadClient` on the IM engine.

So the only thing "shut down subscription" does that just deleting the ReadClient does not is call `OnDone`.  And if we are inside the callback's destructor anyway, there's no point calling OnDone on that callback (and if fact, it would be pretty bad to make virtual function calls on a partially-destroyed object)....

Reply:
if we are inside the callback's destructor anyway, but we do not call ReadClient's Close, mExchange.Release() would not be invoked, chance for crash in exchange layer?

Reply:
> mExchange.Release() would not be invoked

`mExchange` is an `ExchangeHolder` and `~ExchangeHolder` calls `Release`.  We only need the manual release in `Close` in case our callback does not immediately delete our `ReadClient`.  If we are being deleted, then the `Release` happens at that point.